### PR TITLE
Login, Authen and Autho

### PR DESCRIPTION
**Pull request by linhne94**

### CODE REVIEW

File: src/modules/auth/auth.module.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import { AuthController } from './auth.controller';
+import { TypeOrmModule } from '@nestjs/typeorm';
+import { User } from 'src/entities/user.entity';
+import { JwtModule } from '@nestjs/jwt';
+
+@Module({
+  imports: [
+    TypeOrmModule.forFeature([User]),
+    JwtModule.register({
+      secret: 'your_secret_key',
```
Comment (DangHuan2k3) [inline src/modules/auth/auth.module.ts]: move secret jwt to env

Comment (linhne94) [reply src/modules/auth/auth.module.ts]: done

File: src/modules/auth/auth.module.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import { TypeOrmModule } from '@nestjs/typeorm';
+import { User } from 'src/entities/user.entity';
+import { JwtModule } from '@nestjs/jwt';
+
+@Module({
+  imports: [
+    TypeOrmModule.forFeature([User]),
+    JwtModule.register({
+      secret: process.env.JWT_SECRET_KEY,
+      signOptions: { expiresIn: '15d' },
```
Comment (AnhNN11) [inline src/modules/auth/auth.module.ts]: config expiresIn: '15d'  into file .env

File: src/modules/auth/auth.service.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+@Injectable()
+export class AuthService {
+
+  constructor(
+    @InjectRepository(User)
+    private readonly userRepository: Repository<User>,
+    private readonly jwtService: JwtService,
+  ) {}
+
+  async login(loginDto: LoginDto): Promise<{ token: string }> {
```
Comment (AnhNN11) [inline src/modules/auth/auth.service.ts]: try-catch login

File: src/guards/jwt-auth.guard.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import { Injectable, CanActivate, ExecutionContext } from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { JwtService } from '@nestjs/jwt';
+
+@Injectable()
+export class JwtAuthGuard implements CanActivate {
+  constructor(private reflector: Reflector, private jwtService: JwtService) {}
+
+  canActivate(context: ExecutionContext): boolean {
```
Comment (AnhNN11) [inline src/guards/jwt-auth.guard.ts]: async-await và viết exception cụ thể hơn về UnauthorizedException không trả về true false

File: src/guards/roles.guard.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { Injectable, CanActivate, ExecutionContext } from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { UserRole } from 'src/entities/role.enum';
+
+
+@Injectable()
+export class RolesGuard implements CanActivate {
+  constructor(private reflector: Reflector) {}
+
+  canActivate(context: ExecutionContext): boolean {
```
Comment (AnhNN11) [inline src/guards/roles.guard.ts:12]: async-await và viết exception cụ thể hơn về UnauthorizedException không trả về true false

File: src/modules/auth/auth.service.ts [hunk 6]
```diff
@@ -18,15 +18,20 @@ export class AuthService {
   ) {}
 
   async login(loginDto: LoginDto): Promise<{ token: string }> {
-    const user = await this.userRepository.findOneBy({ email: loginDto.email });
-    if(!user){
-      throw new Error('User not found')
+    try {
+      const user = await this.userRepository.findOneBy({ email: loginDto.email });
+      if (!user) {
+        throw new Error('User not found');
```
Comment (AnhNN11) [inline src/modules/auth/auth.service.ts]: throw new BadRequestException({
        message: ERRORS_DICTIONARY. USER_NOT_FOUND
});
You had better only modify the error-dictionary.constraint.ts file in the constraints folder. Avoid hardcoding.

File: src/modules/auth/auth.service.ts [hunk 7]
```diff
@@ -18,15 +18,20 @@ export class AuthService {
...
+      }
+      if (!(await bcrypt.compare(loginDto.password, user.password))) {
+        throw new Error('The password is incorrect');
+      }
+      const payload = { userId: user.id, role: user.role };
+      const token = this.jwtService.sign(payload);
+      return { token };
+    } catch (error) {
+      throw new Error(error.message || 'Login failed');
```
Comment (AnhNN11) [inline src/modules/auth/auth.service.ts]: As mentioned above

File: src/app.module.ts
Comment (NhatHuynh-Madison) [inline src/app.module.ts:1]: conflict indent space (2 space)

File: README.md [hunk 8]
```diff
@@ -83,3 +83,9 @@ Nest is an MIT-licensed open source project. It can grow thanks to the sponsors
 ## License
 
 Nest is [MIT licensed](https://github.com/nestjs/nest/blob/master/LICENSE).
+
+nest g res login
+
+## use Authorize cho các chức năng
+- Ae dùng annotation @UseGuards(JwtAuthGuard, RolesGuard) này ở trên class ...Controller
+- và dùng annotation @Roles('admin') vào chỗ function, route ( ví dụ muốn cho ai cũng dùng được thì đừng thêm @Roles ở trên là được)
```
Comment (Huy-Phan-159) [inline README.md]: please use EN 😄

File: src/modules/auth/auth.controller.ts
Comment (NhatHuynh-Madison) [inline src/modules/auth/auth.controller.ts:1]: add this api to swagger

File: src/modules/auth/auth.service.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  ) {}
+
+  async login(loginDto: LoginDto): Promise<{ token: string }> {
+    try {
+      const user = await this.userRepository.findOneBy({ email: loginDto.email });
+      if (!user) {
+        throw new Error('User not found');
+      }
+      if (!(await bcrypt.compare(loginDto.password, user.password))) {
+        throw new Error('The password is incorrect');
```
Comment (NhatHuynh-Madison) [inline src/modules/auth/auth.service.ts]: follow through BadRequest also

File: src/guards/roles.guard.ts (hunk 5 above)
Comment (NhatHuynh-Madison) [reply src/guards/roles.guard.ts:12]: return true is acceptable, just not return false

File: src/guards/jwt-auth.guard.ts (hunk 4 above)
Comment (NhatHuynh-Madison) [reply src/guards/jwt-auth.guard.ts]: return true is acceptable, just not return false

File: src/modules/auth/auth.controller.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+import { UpdateAuthDto } from './dto/update-auth.dto';
+import { LoginDto } from './dto/login.dto';
+
+@Controller('auth')
+export class AuthController {
+  constructor(private readonly authService: AuthService) {}
+
+  @Post('login')
+  async login(@Body() loginDto: LoginDto) {
+    return this.authService.login(loginDto);
```
Comment (NhatHuynh-Madison) [inline src/modules/auth/auth.controller.ts]: should create a common response object (must work with team to define this response, that will be used through out the project). don't just return data from login for example.

File: src/entities/role.enum.ts [hunk 11]
```diff
@@ -0,0 +1,5 @@
+export enum UserRole {
+    Admin = 'admin',
+    Teacher = 'teacher',
+    Student = 'student',
+  }
```
Comment (Huy-Phan-159) [inline src/entities/role.enum.ts]: Can we reuse the `Role` enum defined here `src/entities/enums/role.enum.ts`?

File: src/guards/jwt-auth.guard.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    const request = context.switchToHttp().getRequest();
+    const token = this.extractTokenFromHeader(request);
+    if (!token) {
+      return false;
```
Comment (Huy-Phan-159) [inline src/guards/jwt-auth.guard.ts]: Should throw `UnauthorizedException` here

File: src/modules/auth/auth.service.ts (hunk 9 above)
Comment (Huy-Phan-159) [reply src/modules/auth/auth.service.ts]: Should return a common message like `The username/password is incorrect` for security purposes

File: src/guards/roles.guard.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+      context.getClass(),
+    ]);
+    if (!requiredRoles) {
+      return true;
+    }
+    const { user } = context.switchToHttp().getRequest();
+    const hasRole = requiredRoles.some((role) => user.role === role);
+
+    if (!hasRole) {
+      throw new BadRequestException(ERRORS_DICTIONARY.AUTHORIZE_ERROR);
```
Comment (Huy-Phan-159) [inline src/guards/roles.guard.ts]: Should throw `ForbiddenException` here

File: src/guards/jwt-auth.guard.ts [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    }
+    try {
+      const payload = this.jwtService.verify(token);
+      // cái request ni sẽ được đem qua roles.guard hắn so sánh
+      request.user = payload;
```
Comment (Huy-Phan-159) [inline src/guards/jwt-auth.guard.ts]: Need to check if user Id is valid or not

**Review by Huy-Phan-159 — CHANGES_REQUESTED** [review]

**Review by Huy-Phan-159 — APPROVED** [review]
LGTM 🚀 🚀

**Review by NhatHuynh-Madison — APPROVED** [review]

**Pull request merged**